src/CloudNative.CloudEvents.Amqp/AmqpCloudEventMessage.cs

```diff
@@ -1,90 +1,96 @@
+                    // AMQPNetLite doesn't support DateTimeOffset values, so convert to UTC.
+                    // That means we can't roundtrip events with non-UTC timestamps, but that's not awful.
+                    DateTimeOffset dto => dto.UtcDateTime,
+                    _ => pair.Value
+                };
+                properties.Add(propKey, propValue);
             }
-            else if (cloudEvent.Data is string)
-            {
-                this.BodySection = new AmqpValue() { Value = cloudEvent.Data };
-            }
-
-            this.Properties = new Properties() { ContentType = cloudEvent.DataContentType?.MediaType };
-            this.ApplicationProperties = new ApplicationProperties();
-            MapHeaders(cloudEvent);
         }
 
-        void MapHeaders(CloudEvent cloudEvent)
+        /// <summary>
+        /// Convert data into a suitable format for inclusion in an Avro record.
+        /// TODO: Asynchronous version?
+        /// </summary>
+        /// <param name="data"></param>
+        /// <returns></returns>
+        private static RestrictedDescribed SerializeData(object data)
         {
-            foreach (var attribute in cloudEvent.GetAttributes())
+            switch (data)
             {
-                if (!attribute.Key.Equals(CloudEventAttributes.DataAttributeName(cloudEvent.SpecVersion)) &&
-                    !attribute.Key.Equals(CloudEventAttributes.DataContentTypeAttributeName(cloudEvent.SpecVersion)))
-                {
-                    string key = "cloudEvents:" + attribute.Key;
-                    if (attribute.Value is Uri)
-                    {
-                        this.ApplicationProperties.Map.Add(key, attribute.Value.ToString());
-                    }
-                    else if (attribute.Value is DateTimeOffset dto)
-                    {
-                        // AMQPNetLite doesn't support DateTimeOffset values, so convert to UTC.
-                        // That means we can't roundtrip events with non-UTC timestamps, but that's not awful.
-                        this.ApplicationProperties.Map.Add(key, dto.UtcDateTime);
-                    }
-                    else if (attribute.Value is string)
-                    {
-                        this.ApplicationProperties.Map.Add(key, attribute.Value);
-                    }
-                    else
-                    {
-                        Map dict = new Map();
-                        foreach (PropertyDescriptor descriptor in TypeDescriptor.GetProperties(attribute.Value))
-                        {
-                            dict.Add(descriptor.Name, descriptor.GetValue(attribute.Value));
-                        }
-
-                        this.ApplicationProperties.Map.Add("cloudEvents:" + attribute.Key, dict);
-                    }
-                }
+                case null:
+                    return null;
+                case byte[] bytes:
```

Comment:
Do we need handling for `ReadOnlyMemory<byte>`?

Reply:
That would require a dependency that I'd rather not take if we can avoid it. But we should definitely think about that when we revisit the question of how we want to handle data in general. (A lot of the data-handling code should be centralized I suspect, in the same way that the attribute-handling code is centralized. There's no need to have multiple copies of "copy a stream into a byte array" code for example.) Definitely raise `ReadOnlyMemory<byte>` again later though :)